File: python/crypto-square/crypto_square.py
import re
import math

def cipher_text(plain_text):
    normalize_plaintext = re.findall(r'\w', plain_text.lower())
    if len(normalize_plaintext) == 0:
        return ''
    col_len = calculate_cols(normalize_plaintext)
    plaintext_segments = []
    for i in range(0, len(normalize_plaintext), col_len):
        plaintext_segments.append(normalize_plaintext[i:i+col_len])

    cipher_segments = [ [] for i in range(col_len)]
    for i in range(0, col_len):
        for row in plaintext_segments:
            if i < len(row):
                cipher_segments[i].append(row[i])
            else:
                cipher_segments[i].append(' ')

    return ' '.join(map(lambda x: ''.join(x), cipher_segments))


def calculate_cols(text):
    col_len = math.floor(math.sqrt(len(text)))
    row_len = len(text) // col_len
    if col_len * row_len < len(text):
        col_len += 1
    if col_len - row_len > 1:
        col_len -= 1
    if col_len < row_len:
        col_len += 1

    return col_len

File: ruby/crypto-square/crypto_square.rb
class Crypto

  attr_reader :normalize_plaintext

  class InvalidColAndRow < StandardError ; end

  def initialize(orig_text)
    @orig_text = orig_text
    @normalize_plaintext = orig_text.scan(/\w/).join.downcase
  end

  def size
    text_length = @normalize_plaintext.length
    cols = Math.sqrt(text_length).ceil
    @rows = text_length / cols
    @rows = @rows + 1 if @rows * cols <= text_length
    raise InvalidColAndRow if cols - @rows > 1
    cols
  end

  def plaintext_segments
    @normalize_plaintext.chars.each_slice(size).reduce([]) {|sum, e| sum << e.join}
  end

  def ciphertext
    ciphertext_ = []
    size.times { |n| ciphertext_ = ciphertext_ + plaintext_segments.map { |e| e.chars[n] } }
    ciphertext_.join
  end

  def normalize_ciphertext
    normalize_ciphertext_ = []
    size.times { |n| normalize_ciphertext_ << plaintext_segments.map { |e| e.chars[n].nil? ? ' ' : e.chars[n] }.join.strip }
    normalize_ciphertext_.join(' ').strip
  end

end
